From c64a6c68a536417bfa81adfad9b6641aff1b128e Mon Sep 17 00:00:00 2001
From: Piotr Gorski <lucjan.lucjanov@gmail.com>
Date: Fri, 20 May 2022 09:10:30 +0200
Subject: [PATCH 2/3] Revert "blk-mq: don't queue 'hctx->run_work' if the queue
 is dead"

This reverts commit fa0060bb513efab194fc528e693d44a3ff9c19d0.

Signed-off-by: Piotr Gorski <lucjan.lucjanov@gmail.com>
---
 block/blk-core.c |  3 ---
 block/blk-mq.c   | 10 ++--------
 2 files changed, 2 insertions(+), 11 deletions(-)

diff --git a/block/blk-core.c b/block/blk-core.c
index 2cdfb9bf3..bc0506772 100644
--- a/block/blk-core.c
+++ b/block/blk-core.c
@@ -314,10 +314,7 @@ void blk_cleanup_queue(struct request_queue *q)
 	 */
 	blk_freeze_queue(q);
 
-	/* New 'hctx->run_work' can't be queued after setting the dead flag */
-	spin_lock_irq(&q->queue_lock);
 	blk_queue_flag_set(QUEUE_FLAG_DEAD, q);
-	spin_unlock_irq(&q->queue_lock);
 
 	blk_sync_queue(q);
 	if (queue_is_mq(q)) {
diff --git a/block/blk-mq.c b/block/blk-mq.c
index 3e9bf2eb2..84d749511 100644
--- a/block/blk-mq.c
+++ b/block/blk-mq.c
@@ -2042,8 +2042,6 @@ static int blk_mq_hctx_next_cpu(struct blk_mq_hw_ctx *hctx)
 static void __blk_mq_delay_run_hw_queue(struct blk_mq_hw_ctx *hctx, bool async,
 					unsigned long msecs)
 {
-	unsigned long flag;
-
 	if (unlikely(blk_mq_hctx_stopped(hctx)))
 		return;
 
@@ -2058,12 +2056,8 @@ static void __blk_mq_delay_run_hw_queue(struct blk_mq_hw_ctx *hctx, bool async,
 		put_cpu();
 	}
 
-	spin_lock_irqsave(&hctx->queue->queue_lock, flag);
-	if (!blk_queue_dead(hctx->queue))
-		kblockd_mod_delayed_work_on(blk_mq_hctx_next_cpu(hctx),
-					    &hctx->run_work,
-					    msecs_to_jiffies(msecs));
-	spin_unlock_irqrestore(&hctx->queue->queue_lock, flag);
+	kblockd_mod_delayed_work_on(blk_mq_hctx_next_cpu(hctx), &hctx->run_work,
+				    msecs_to_jiffies(msecs));
 }
 
 /**
-- 
2.36.1.74.g277cf0bc36

